bigtable/client/table.h

```diff
@@ -220,6 +224,59 @@ class Table {
                          std::vector<Mutation> true_mutations,
                          std::vector<Mutation> false_mutations);
 
+  using RpcUtils = bigtable::internal::UnaryRpcUtils<DataClient>;
+  using StubType = RpcUtils::StubType;
+
+  /**
+   * Atomically read and modify the row in the server, returning the
+   * resulting row
+   *
+   * @tparam Args this is zero or more ReadModifyWriteRules to apply on a row
+   * @param row_key the row to read
+   * @param rule to modify the row. Two types of rules are applied here
+   *     AppendValue which will read the existing value and append the
+   *     text provided to the value.
+   *     IncrementAmount which will read the existing uint64 big-endian-int
+   *     and add the value provided.
+   *     Both rules accept the family and column identifier to modify.
+   * @param rules is the zero or more ReadModifyWriteRules to apply on a row.
+   * @returns modified row
+   */
+  template <typename... Args>
+  Row ReadModifyWriteRow(std::string row_key,
+                         bigtable::ReadModifyWriteRule rule, Args&&... rules) {
+    btproto::ReadModifyWriteRowRequest request;
+    request.set_table_name(table_name_);
+    request.set_row_key(std::move(row_key));
+
+    // Generate a better compile time error message than the default one
+    // if the types do not match
+    static_assert(
+        internal::conjunction<
+            std::is_convertible<Args, bigtable::ReadModifyWriteRule>...>::value,
+        "The arguments passed to ReadModifyWriteRow(row_key,...) must be "
+        "convertible to bigtable::ReadModifyWriteRule");
+
+    // Below code to add the ReadModifyWriteRule to the request object
+    // is by making a copy. This will create performance problem if the
+    // rules buffer is large
+    // TODO Issue#336
```

Comment:
We have been using `// TODO(#336) - optimize this code by not copying the parameter pack`.